### PR TITLE
proposal(dot/rpc/subscription): refactoring the websocket `HandleComm`

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Adding `getSetupListener` and `getUnsubListener` to avoid code duplication, these functions have internal switches to get the right function to the right method.
- Using mutexes when updating the `qtyListeners`.
- Adding `Stop` to the `Listener` and add `ctx` and `cancel` to the listeners' implementation, this way is possible to stop the inners goroutines when any `unsubscriber` call is made.

## Tests

```
go test -run ^TestWSConn_HandleComm$ github.com/ChainSafe/gossamer/dot/rpc/subscription
```

## Issues

- N/A

### CODE REVIEW

File: dot/rpc/subscription/listeners.go [hunk 1]
```diff
@@ -85,59 +87,90 @@ func (s *StorageObserver) GetFilter() map[string][]byte {
...
+	l.ctx, l.cancel = context.WithCancel(context.Background())
+	go func() {
+		for {
+			select {
+			case <-l.ctx.Done():
+				return
+			case block := <-l.Channel:
+				if block == nil {
+					continue
+				}
```
Comment (noot) [inline dot/rpc/subscription/listeners.go:115]: when does the channel get closed?

check if channel is closed:
```suggestion
			case block, ok := <-l.Channel:
				if !ok {
					return
				}
```

File: dot/rpc/subscription/websocket.go [hunk 2]
```diff
@@ -51,201 +57,165 @@ type WSConn struct {
...
 	RPCHost            string
+
+	HTTP httpclient
+}
+
+// readWebsocketMessage will read and parse the message data to a string->interface{} data
+func (c *WSConn) readWebsocketMessage() ([]byte, map[string]interface{}, error) {
+	_, mbytes, err := c.Wsconn.ReadMessage()
+	if err != nil {
+		logger.Warn("websocket failed to read message", "error", err)
```
Comment (noot) [inline dot/rpc/subscription/websocket.go:68]: I'd probably put this as debug

File: dot/rpc/subscription/websocket.go [hunk 3]
```diff
@@ -51,201 +57,165 @@ type WSConn struct {
...
+	}
+
+	logger.Trace("websocket received", "message", mbytes)
+
+	// determine if request is for subscribe method type
+	var msg map[string]interface{}
+	err = json.Unmarshal(mbytes, &msg)
+
+	if err != nil {
+		logger.Warn("websocket failed to unmarshal request message", "error", err)
```
Comment (noot) [inline dot/rpc/subscription/websocket.go:79]: same here

File: dot/rpc/subscription/websocket.go [hunk 4]
```diff
@@ -256,61 +226,75 @@ func (c *WSConn) unsubscribeStorageListener(reqID float64, params interface{}) {
 	c.safeSend(newBooleanResponseJSON(true, reqID))
 }
 
-func (c *WSConn) initBlockListener(reqID float64) (uint, error) {
+func (c *WSConn) initBlockListener(reqID float64, params interface{}) (Listener, error) {
```
Comment (noot) [inline dot/rpc/subscription/websocket.go:229]:
```suggestion
func (c *WSConn) initBlockListener(reqID float64, _ interface{}) (Listener, error) {
```

File: dot/rpc/subscription/websocket.go [hunk 5]
```diff
@@ -256,61 +226,75 @@ func (c *WSConn) unsubscribeStorageListener(reqID float64, params interface{}) {
...
+	c.mu.Unlock()
+
+	c.safeSend(NewSubscriptionResponseJSON(bl.subID, reqID))
 
-	return bl.subID, nil
+	return bl, nil
 }
 
-func (c *WSConn) initBlockFinalizedListener(reqID float64) (uint, error) {
+func (c *WSConn) initBlockFinalizedListener(reqID float64, params interface{}) (Listener, error) {
```
Comment (noot) [inline dot/rpc/subscription/websocket.go:261]:
```suggestion
func (c *WSConn) initBlockFinalizedListener(reqID float64, _ interface{}) (Listener, error) {
```

File: dot/rpc/subscription/websocket.go [hunk 6]
```diff
@@ -350,24 +338,30 @@ func (c *WSConn) initExtrinsicWatch(reqID float64, params interface{}) (uint, er
 	c.safeSend(newSubscriptionResponse(AuthorExtrinsicUpdates, esl.subID, "ready"))
 
 	// todo (ed) determine which peer extrinsic has been broadcast to, and set status
-	return esl.subID, err
+	return esl, err
 }
 
-func (c *WSConn) initRuntimeVersionListener(reqID float64) (uint, error) {
+func (c *WSConn) initRuntimeVersionListener(reqID float64, params interface{}) (Listener, error) {
```
Comment (noot) [inline dot/rpc/subscription/websocket.go:344]:
```suggestion
func (c *WSConn) initRuntimeVersionListener(reqID float64, _ interface{}) (Listener, error) {
```

**Review by noot — APPROVED** [review]
looks good - only minor comments

File: dot/rpc/subscription/listeners.go (hunk 1 above)
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/listeners.go:115]: Check added!

File: dot/rpc/subscription/websocket.go (hunk 2 above)
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go:68]: Done!

File: dot/rpc/subscription/websocket.go (hunk 3 above)
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go:79]: Done!

File: dot/rpc/subscription/subscription.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+		return c.initBlockFinalizedListener
+	case "state_subscribeRuntimeVersion":
+		return c.initRuntimeVersionListener
+	default:
+		return nil
+	}
+}
+
+func (c *WSConn) getUnsubListener(method string, params interface{}) (unsubListener, Listener, error) {
+	subscriberid, err := parseSubscribeID(params)
```
Comment (arijitAD) [inline dot/rpc/subscription/subscription.go:33]: `subscriberid` -> `subscriberID`

File: dot/rpc/subscription/websocket.go [hunk 8]
```diff
@@ -395,6 +390,53 @@ func (c *WSConn) safeSendError(reqID float64, errorCode *big.Int, message string
 	}
 }
 
+func (c *WSConn) prepareRequest(b []byte) (*http.Request, error) {
+	buff := &bytes.Buffer{}
+	_, err := buff.Write(b)
+	if err != nil {
```
Comment (arijitAD) [inline dot/rpc/subscription/websocket.go:396]:
```
if _, err := buff.Write(b); err != nil {
```

File: dot/rpc/subscription/websocket.go [hunk 9]
```diff
@@ -408,6 +450,6 @@ type ErrorMessageJSON struct {
 	Message string   `json:"message"`
 }
 
-func (c *WSConn) startListener(lid uint) {
-	go c.Subscriptions[lid].Listen()
+func contains(s string, sub string) bool {
```
Comment (arijitAD) [inline dot/rpc/subscription/websocket.go:453]: Why is this wrapper needed?
We can directly check for
```
strings.Contains(s, sub)
```

**Review by arijitAD — APPROVED** [review]

**Review by edwardmack — APPROVED** [review]
Great work, lgtm!

File: dot/rpc/subscription/websocket_test.go [hunk 10]
```diff
@@ -127,6 +138,7 @@ func TestWSConn_HandleComm(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, []byte(`{"jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request"},"id":7}`+"\n"), msg)
 
+	fmt.Println(wsconn.Subscriptions)
```
[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket_test.go:141]: done!

**Pull request merged**